### PR TITLE
Add interceptor to create noop spans when sampling is false

### PR DESCRIPTION
**Pull request by objectiser**
To avoid unnecessary performance overhead (generating random span ids) when a request to the mixer is not being sampled, this PR will use a `NoopSpan` when the `B3` headers indicate that the request is not sampled.

@douglas-reid @mandarjog Submitted for initial comments - plan to add some more tests using the mock tracer.

Supercedes #10569 which was the initial prototype work by @douglas-reid

### CODE REVIEW

**Comment by douglas-reid** [comment]
This looks good to me. Thanks for taking this on.

One concern: there are a bunch of changes in the dep lock file. Did you use the appropriate version of `dep`? It feels like the changes are related to using a newer/older version than was previously used.

**Comment by yurishkuro** [comment]
@objectiser this makes sense for a leaf service in the call graph, but for a node in the middle the noop span would result in no trace context sent downstream, which may cause the downstream service to start a new trace. Is that what you intended?

**Comment by objectiser** [comment]
@douglas-reid Looks like the dep version was wrong - updated and it looks better now.

**Comment by objectiser** [comment]
@douglas-reid Any ideas why the lint task is failing? I've ran `make format` and it didn't make any changes - and the file listed at the end of the log `mixer/pkg/runtime/dispatcher/tracing_test.go:8: File is not `goimports`-ed (goimports)` hasn't been changed in this PR.

File: mixer/pkg/server/interceptor.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		if !ok {
+			md = metadata.New(nil)
+		}
+
+		var t opentracing.Tracer = defaultNoopTracer
+		if isSampled(md) {
+			t = tracer
+		}
+
+		spanContext, err := t.Extract(opentracing.HTTPHeaders, metadataReaderWriter{md})
```
Comment (douglas-reid) [inline mixer/pkg/server/interceptor.go]: in the other PR, we were checking to see if this `err` was `opentracing.ErrSpanContextNotFound`. This, I think, was forcing noopSpans for cases when no trace headers were forwarded (which I believe is still the desired behavior). Should something like that be included here?

File: mixer/pkg/server/interceptor_test.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	info := &grpc.UnaryServerInfo{
+		FullMethod: "mymethod",
+	}
+	interceptor(ctx, nil, info, func(ctx context.Context, req interface{}) (interface{}, error) {
+		return nil, nil
+	})
+	assert.Len(t, tracer.FinishedSpans(), 1)
+}
+
+func TestUnampledSpan(t *testing.T) {
```
Comment (douglas-reid) [inline mixer/pkg/server/interceptor_test.go]: nit: missing 's' between 'Un' and 'ampled'

**Comment by douglas-reid** [comment]
@objectiser I've seen failures in other PRs related to `mixer/pkg/runtime/dispatcher/tracing_test.go`. I don't know if it has been fixed in another PR and would be patched by a rebase or if the linter is in a funny state, however.

**Comment by objectiser** [comment]
@douglas-reid Updated to take `opentracing.ErrSpanContextNotFound` into account.

**Comment by douglas-reid** [comment]
/lgtm

New linter error: 

```
WARN [runner/nolint] Found unknown linters in //nolint directives: vet, vetshadow 
mixer/pkg/server/interceptor.go:63:9: should omit type opentracing.Tracer from declaration of var t; it will be inferred from the right-hand side (golint)
		var t opentracing.Tracer = tracer
```

**Comment by objectiser** [comment]
@douglas-reid Sorry hadn't checked. Now is back to the same lint error as before.

**Comment by douglas-reid** [comment]
@objectiser do you want to just fix it in this PR?  If not, I can take a stab at a small PR to fix.

**Comment by objectiser** [comment]
@douglas-reid Separate PR if you don't mind, unless you can let me know how to fix - I tried running `make format` and it didn't fit the problem, so wasn't sure how to fix.

**Comment by objectiser** [comment]
@douglas-reid Not sure what is different about release-1.1 branch, but I've tested an earlier version of the changes from master branch, but when I build, `make installgen`, create the helm chart (with tracing enabled) and deploy to minikube, I get the following errors:

```
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "attributemanifest" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "attributemanifest" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "handler" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "logentry" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "logentry" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "metric" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "handler" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "handler" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "rule" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "kubernetes" in version "config.istio.io/v1alpha2"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "DestinationRule" in version "networking.istio.io/v1alpha3"
unable to recognize "/home/gbrown/istio-test.yaml": no matches for kind "DestinationRule" in version "networking.istio.io/v1alpha3"
```
Any ideas?

**Comment by douglas-reid** [comment]
@objectiser i ran into this as well. you now need to install the istio-init crd yamls first (via `kubectl apply -f install/kubernetes/helm/istio-init/files/` or similar).

**Comment by douglas-reid** [comment]
@objectiser have you forced pushed a rebase to 1.1 tip recently by any chance? This PR (https://github.com/istio/istio/pull/10831) should have fixed the issue you are seeing with the linter.

**Comment by objectiser** [comment]
@douglas-reid Thanks that seems to have sorted it.

**Comment by objectiser** [comment]
@douglas-reid Needs further testing - no getting any spans reported from the mixer when doing test with bookinfo. As mentioned a test on master branch with an earlier version of the change worked - so will need to investigate what is different.

**Comment by objectiser** [comment]
@douglas-reid Tested out the PR on master and works fine (istio-mixer spans shown in trace) - also tested the PR on release-1.1 branch with some additional diagnostics that show working correctly. However on rel-1.1 branch the istio-mixer spans are not being reported (i.e. to the backend)

I did a quick test on the latest rel-1.1 HEAD and it also doesn't report istio-mixer spans - not sure if a change has recently been made to disable them by default? I can dig further next week - but just in case you know off the top of your head?

So basically I believe the PR is working fine - just need to confirm why istio-mixer spans are not being reported.

**Comment by douglas-reid** [comment]
@objectiser there was a PR awhile back that turned off the tracing generation at the proxy that sits in front of Mixer, meaning that only the Mixer process itself should generate traces, but that should not have disabled Mixer tracing altogether (unless that means that envoy drops the headers for some reason??).

I can take a look. In the meantime, can you generate trace data with `mixs` + this change?

**Comment by objectiser** [comment]
@douglas-reid From the diagnostics, it showed the trace interceptor was receiving a valid trace context - so the doesn't look like the headers are being dropped - just the spans are not being reported for some reason.

Sorry don't know what `mixs` is?

**Comment by douglas-reid** [comment]
@objectiser `mixs` is the mixer process. you can `make mixs` and build the client too (`make mixc`) to do local testing of mixer.

I tested with `gcr.io/istio-release/mixer:release-1.1-latest-daily` on a cluster just now. I used fortio to send 1000 requests, and was able to see the following. I think you might be seeing the effectiveness of caching in this scenario.

trace-istio-mixer-release1 1
